### PR TITLE
Base idea for resource requirement APIs

### PR DESCRIPTION
**Pull request by nicelhc13**
@spawn(placement=[cuda(0)[{"vcu": 10, "memory":10}], ..)
is the plan. still work in progress.

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -11,26 +11,30 @@
...
+    """
+    @spawn(T[0])
     def task1():
         print("+HELLO OUTER 0", flush=True)
         bsleep(1000)
         print("-HELLO OUTER 0", flush=True)
 
-    @spawn(T[1], placement=[cpu], vcus=0, dependencies=[T[0]])
+    """
+    @spawn(T[0], placement=[(cpu[{"vcus":20, "memory":40000}], cuda[{"vcus":1, "memory":20}]), (cpu[0], cuda[1][{"vcus":10, "memory":2000}])])
```
Comment (dialecticDolt) [inline test.py:22]: - I really like get_item instead of () for specifying which device. Especially because it would allow us to do slices to specify a multi-device tuple... although then it is unclear how to specify constraints. 

- Why `get_item` for both the constraint and which device? (this is inconsistent with the original PR comment). Is this possibly confusing for users? What should the edge cases of specifying multiple get_items resolve to cuda[1][2], cuda[{'memory': 100}]['{memory': 100, 'vcu': 1}]?

Comment (nicelhc13) [reply test.py:22]: yeah that was also my confusion but just left that at that moment. I just followed the old Parla just in case that there is another reason that idk. So, if you also thought like me, let me remove it. i don't understand the latter example. Is this an existing edge case? we should provide a resource requirement for a single device or for a single architecture. I don't understand what cuda[1][2] means and also the latter one.
